### ISSUE TITLE
Digital Music Packs

### ISSUE THREAD

**Issue by thenodger** [fabiangreffrath/woof#1979]
I am having trouble getting the [Roland SC-55 music packs](http://sc55.duke4.net/games.php#doom) working well with Woof. When I load them using the autoload folders, the TNT music just plays when playing Doom 2 style mapXYs. Using Crispy Doom, I was able to get this working flawlessly: I made a config file for each of Doom, Doom II and TNT, and when playing Plutonia, the correct track would be selected out of the three music packs. How can I get it working like this with Woof?

**Comment by rfomin** [comment]
How do you use autoload folders? If you put "tntevilution_sc55_ogg.zip" in the autoload/tnt.wad folder, it should only work for TNT.WAD.

**Comment by thenodger** [comment]
> How do you use autoload folders? If you put "tntevilution_sc55_ogg.zip" in the autoload/tnt.wad folder, it should only work for TNT.WAD.

Yup, that works for TNT, but still, it doesn't make the solution I want. If a WAD needs music from TNT, it won't be able to get it, and now Plutonia just plays the Doom 2 music rather than the correct music for each level.

**Comment by fabiangreffrath** [comment]
Sorry, but I still don't get your problem. If you load a PWAD with Doom 2 as the IWAD, it cannot somehow "need music from TNT". Doom 2, TNT and Plutonia all use the same lump names for their music.

**Comment by fabiangreffrath** [comment]
Wait a minute. Is this what you're looking for?

https://github.com/fabiangreffrath/crispy-doom/blob/5fe53320ae3e07a106c27b7b46282d8efc490378/src/doom/s_sound.c#L112-L233

**Comment by ceski-1** [comment]
> How can I get it working like this with Woof?

There's a script on that page that generates the Plutonia music pack from the Doom 1/2 music packs. The files are placed in a temporary folder like this:

```
7za.exe
doom2_sc55_ogg_v1.2.zip
doom_sc55_ogg.zip
plutoniamusic.bat
```

Then run `plutoniamusic.bat` and it will output `plutonia_sc55_ogg.zip`. After that, place the zip files in the appropriate autoload folders:

```
woof
  autoload
    doom.wad
      doom_sc55_ogg.zip
    doom2.wad
      doom2_sc55_ogg_v1.2.zip
    plutonia.wad
      plutonia_sc55_ogg.zip
    tnt.wad
      tntevilution_sc55_ogg.zip
```

Works fine for me. There's no need for music configs.

**Comment by thenodger** [comment]
> Sorry, but I still don't get your problem. If you load a PWAD with Doom 2 as the IWAD, it cannot somehow "need music from TNT". Doom 2, TNT and Plutonia all use the same lump names for their music.

I thought that some PWADs would use TNT MIDI tracks, but I guess that's not a thing then.


> Wait a minute. Is this what you're looking for?
> 
> https://github.com/fabiangreffrath/crispy-doom/blob/5fe53320ae3e07a106c27b7b46282d8efc490378/src/doom/s_sound.c#L112-L233

I didn't actually use that method, because the new TNT tracks had the same filenames as the ones from Doom II, I used the config file method instead, which is marked as obsolete: [https://github.com/chocolate-doom/music-packs](https://github.com/chocolate-doom/music-packs).
Plutonia and No Rest for the Living were able to just take the correct tracks without the need to create a seperate pack for Plutonia.



> Sorry, but I still don't get your problem.

OK, so what I want is for the digital music tracks to just be used when they match the MIDI of the current level. Currently, loading the music packs just plays them in order for the IWAD they are for, regardless of what MIDI the level plays; this works for the IWADs, and I could just make a music pack for Plutonia, however, it wouldn't work for PWADs, such as NRftL, which plays some of the DOOM II tracks in its own order.

On the subject of NRftL, using nerve.wad acquired from DOOM + DOOM II doesn't work with Woof. It works fine with Crispy Doom, but I get this error on the standard output when using Woof:
`$ /appimages/Woof-14.5.0-Linux.appimage -iwad games/doom/iwads/DOOM2.WAD -file games/doom/maps/nerve.wad`
`Segmentation fault (core dumped)`

**Comment by ceski-1** [comment]
> OK, so what I want is for the digital music tracks to just be used when they match the MIDI of the current level. Currently, loading the music packs just plays them in order for the IWAD they are for, regardless of what MIDI the level plays; this works for the IWADs, and I could just make a music pack for Plutonia, however, it wouldn't work for PWADs, such as NRftL, which plays some of the DOOM II tracks in its own order.

NRftL works as expected:

```
$ ./woof.appimage -verbose -file nerve.wad -cl 2

W_Init: Init WADfiles.
 adding DOOM2.WAD
 adding /home/username/.local/share/woof/autoload/doom2.wad/doom2_sc55_ogg_v1.2.zip
 adding nerve.wad

[...]

S_ChangeMusInfoMusic: D_MESSAG (lump), Ogg Vorbis (Xiph Foundation)
P_SetupLevel: MAP01 (nerve.wad), Skill 3, Doom, Doom 1.9
```

Cycling through the maps matches the track order listed [here](https://doomwiki.org/wiki/No_Rest_for_the_Living).

> On the subject of NRftL, using nerve.wad acquired from DOOM + DOOM II doesn't work with Woof. It works fine with Crispy Doom, but I get this error on the standard output when using Woof: `$ /appimages/Woof-14.5.0-Linux.appimage -iwad games/doom/iwads/DOOM2.WAD -file games/doom/maps/nerve.wad` `Segmentation fault (core dumped)`

This is already fixed for the next release. Try the AppImage at the bottom of [this page](https://github.com/fabiangreffrath/woof/actions/runs/11659011288).

**Comment by thenodger** [comment]
> NRftL works as expected

...
> Cycling through the maps matches the track order listed [here](https://doomwiki.org/wiki/No_Rest_for_the_Living).

Sweet! Since it wasn't working, I assumed that NRftL wouldn't do this and it would just play the Doom II tracks in Doom II's order, like Plutonia does. I guess PWADs work fine then.

> This is already fixed for the next release. Try the AppImage at the bottom of [this page](https://github.com/fabiangreffrath/woof/actions/runs/11659011288).

Yup, this works, thanks! Is stability a concern with this, as it's an unreleased build or whatever you'd call it?

**Comment by ceski-1** [comment]
> Is stability a concern with this, as it's an unreleased build or whatever you'd call it?

They're development builds (GitHub calls them artifacts), so there's always a chance that something isn't stable. If you're okay with that and want to try the latest changes, bookmark [this page](https://github.com/fabiangreffrath/woof/actions/workflows/packages.yml?query=branch%3Amaster) and check back once in a while. Otherwise, official releases will always be listed [here](https://github.com/fabiangreffrath/woof/releases).

Is this issue resolved now?

**Comment by rfomin** [comment]
> All the sourceports come with crappy MIDI soundfonts or just use your OSs MIDI software, but at that point this is a wider Doom community issue rather than anything specific to this sourceport, so I guess I'll close this thread.

We can't bundle good soundfonts due to copyright issues. Consider using the https://github.com/nukeykt/Nuked-SC55/ emulator, not easy to set up, but you will get perfect SC-55 music for any WAD.
Most users are on Windows, which has the Roland SC emulation enabled by default. It's crude, but good enough for Doom music.

**Comment by fabiangreffrath** [comment]
Hm, I still wonder if we could get away with just a single music pack for Doom 2, TNT and Plutonia if we adopted the code in Crispy that I linked to.
